pilot/pkg/xds/discovery.go

```diff
@@ -103,13 +100,18 @@ type DiscoveryServer struct {
 	// the push context, which means that the next push to a proxy will receive this configuration.
 	CommittedUpdates *atomic.Int64
 
+	// mutex used for protecting shards.
+	mutex sync.RWMutex
 	// EndpointShards for a service. This is a global (per-server) list, built from
 	// incremental updates. This is keyed by service and namespace
 	EndpointShardsByService map[string]map[string]*EndpointShards
 
+	// pushChannel is the buffer used for debouncing,
+	// after debouncing the pushRequest will be send to pushQueue
 	pushChannel chan *model.PushRequest
 
-	// mutex used for config update scheduling (former cache update mutex)
+	// mutex used for Environment.PushContext
+	// TODO: is this necessary?
```

Comment:
cc @ramaraochavali

Reply:
Yes it is. At the very least, it is required for caching to properly invalidate (https://github.com/istio/istio/blob/2c583b8567d361753d6be56d62cc63089407c406/pilot/pkg/xds/discovery.go#L523). But also probably required to prevent standard data races.

Note we have failing tests without the line linked

Reply:
there is also rarely a downside to having a mutex, just to be on the safe side.

Reply:
PushContext is different, before each epoch of push, it is reinitiated. So it is worth revisit, so i left it there

Reply:
@howardjohn  why need to protect dropCacheForRequest?  underlying cache has its own lock

Reply:
I believe you are right but seems this is very likely error-prone if someone later touched these part without so much knowledge.